lib/Serialization/Deserialization.cpp

```diff
@@ -2810,37 +2818,42 @@ ModuleFile::getDeclChecked(DeclID DID, Optional<DeclContext *> ForcedContext) {
     bool isStatic;
     uint8_t rawStaticSpelling, rawAccessLevel, rawAddressorKind;
     bool isObjC, isMutating, hasDynamicSelf, throws;
-    unsigned numParamPatterns;
+    unsigned numParamPatterns, numNameComponentsBiased;
     GenericEnvironmentID genericEnvID;
-    TypeID interfaceTypeID, canonicalTypeID;
+    TypeID interfaceTypeID;
     DeclID associatedDeclID;
     DeclID overriddenID;
     DeclID accessorStorageDeclID;
-    bool hasCompoundName, needsNewVTableEntry;
-    ArrayRef<uint64_t> nameIDs;
+    bool needsNewVTableEntry;
+    ArrayRef<uint64_t> nameAndDependencyIDs;
 
     decls_block::FuncLayout::readRecord(scratch, contextID, isImplicit,
                                         isStatic, rawStaticSpelling, isObjC,
                                         isMutating, hasDynamicSelf, throws,
                                         numParamPatterns, genericEnvID,
-                                        interfaceTypeID, canonicalTypeID,
+                                        interfaceTypeID,
                                         associatedDeclID, overriddenID,
-                                        accessorStorageDeclID, hasCompoundName,
+                                        accessorStorageDeclID,
+                                        numNameComponentsBiased,
                                         rawAddressorKind, rawAccessLevel,
-                                        needsNewVTableEntry, nameIDs);
+                                        needsNewVTableEntry,
+                                        nameAndDependencyIDs);
 
     // Resolve the name ids.
-    SmallVector<Identifier, 2> names;
-    for (auto nameID : nameIDs)
-      names.push_back(getIdentifier(nameID));
-
+    Identifier baseName = getIdentifier(nameAndDependencyIDs.front());
     DeclName name;
-    if (!names.empty()) {
-      if (hasCompoundName)
-        name = DeclName(ctx, names[0],
-                        llvm::makeArrayRef(names.begin() + 1, names.end()));
-      else
-        name = DeclName(names[0]);
+    if (numNameComponentsBiased != 0) {
+      SmallVector<Identifier, 2> names;
+      for (auto nameID : nameAndDependencyIDs.slice(1,
+                                                    numNameComponentsBiased-1)){
+        names.push_back(getIdentifier(nameID));
+      }
+      name = DeclName(ctx, baseName, names);
+      nameAndDependencyIDs =
+          nameAndDependencyIDs.slice(numNameComponentsBiased);
+    } else {
+      name = baseName;
+      nameAndDependencyIDs = nameAndDependencyIDs.drop_front();
```

Comment:
The change of meaning of `nameAndDependencyIDs` here really threw me. The code is correct, but kinda hard to read. What about just defining

    auto dependencyIDs = nameAndDependencyIDs.slice(numNameComponentsBiased);

?